File: sdks/python/examples/approval_flow.py
"""Example: Create and respond to an approval request."""

from atap import ATAPClient, ApprovalSubject

# Requester: create an approval
requester = ATAPClient(
    base_url="http://localhost:8080",
    did="did:web:localhost%3A8080:agent:requester",
    private_key="<base64 private key>",
    client_secret="atap_...",
)

approval = requester.approvals.create(
    from_did="did:web:localhost%3A8080:agent:requester",
    to_did="did:web:localhost%3A8080:human:approver",
    subject=ApprovalSubject(
        type="com.example.payment",
        label="Payment of $100 to Acme Corp",
        payload={"amount": 100, "currency": "USD", "recipient": "Acme Corp"},
    ),
)
print(f"Approval created: {approval.id} (state: {approval.state})")

# Approver: respond to the approval
approver = ATAPClient(
    base_url="http://localhost:8080",
    did="did:web:localhost%3A8080:human:approver",
    private_key="<base64 private key>",
)
# For human entities, use authorization_code flow:
approver.token_manager.obtain_authorization_code()

# List pending approvals
pending = approver.approvals.list()
for a in pending:
    print(f"  [{a.state}] {a.id}: {a.subject.label if a.subject else 'N/A'}")

# Approve the first one
if pending:
    result = approver.approvals.respond(pending[0].id, signature="<JWS signature>")
    print(f"Responded: {result.id} -> {result.state}")

requester.close()
approver.close()

File: sdks/javascript/examples/approval-flow.ts
/**
 * Example: Create and respond to approvals.
 *
 * Run with: npx tsx examples/approval-flow.ts
 */

import { ATAPClient } from "@atap/sdk";
import type { ApprovalSubject } from "@atap/sdk";

async function main() {
  // Create an authenticated client
  const client = new ATAPClient({
    baseUrl: "http://localhost:8080",
    did: "did:web:localhost%3A8080:agent:requester",
    privateKey: "<base64 Ed25519 seed>",
    clientSecret: "atap_...",
  });

  // Create an approval request
  const subject: ApprovalSubject = {
    type: "com.example.payment",
    label: "Payment of $100 to vendor",
    payload: {
      amount: 100,
      currency: "USD",
      recipient: "vendor@example.com",
    },
  };

  const approval = await client.approvals.create(
    "did:web:localhost%3A8080:agent:requester",
    "did:web:localhost%3A8080:human:approver",
    subject,
  );

  console.log("Created approval:");
  console.log(`  ID:    ${approval.id}`);
  console.log(`  State: ${approval.state}`);
  console.log(`  From:  ${approval.fromDid}`);
  console.log(`  To:    ${approval.toDid}`);

  // List pending approvals
  const approvals = await client.approvals.list();
  console.log(`\nPending approvals: ${approvals.length}`);

  // Respond to an approval (as the approver)
  const responded = await client.approvals.respond(approval.id, "<JWS signature>");
  console.log(`\nApproval ${responded.id} is now: ${responded.state}`);
}

main().catch(console.error);
